### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by gregorgabriel**
lambda_handler is calling class RequestHandler to handle requests.
two event types are covered: MODIFY and INSERT to be persisted in dynamoDB 
Some validation/error handling for mandatory paramterers.

### CODE REVIEW

File: src/common/encoders.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        type_name = instance.__class__.__name__
+        raise TypeError(f"Object of type {type_name} is not JSON serializable")
+
+
+def remove_none_values(temp_value):
+    return_value = dict()
+    for key, value in temp_value.items():
+        if value is not None:
+            return_value[key] = value
+    return return_value
```
Comment (gregorgabriel) [inline src/common/encoders.py]: same method as in helper.py

File: src/common/helpers.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+def remove_none_values(temp_value):
+    return_value = dict()
+    for key, value in temp_value.items():
+        if value is not None:
+            return_value[key] = value
+    return return_value
```
Comment (gregorgabriel) [inline src/common/helpers.py]: same method as in encoders.py

File: test/test_RequestHandler.py [hunk 3]
```diff
@@ -1,64 +1,100 @@
...
+    body_value = json.dumps(body, default=encode_body)
+    return {
+        'body': body_value
+    }
 
 
 @mock_dynamodb2
 class TestHandlerCase(unittest.TestCase):
+    EXISTING_RESOURCE_IDENTIFIER = 'ebf20333-35a5-4a06-9c58-68ea688a9a8b'
+    EXISTING_RESOURCE_IDENTIFIER_MISSING_CREATED_DATE = 'acf20333-35a5-4a06-9c58-68ea688a9a9c'
```
Comment (gregorgabriel) [inline test/test_RequestHandler.py]: this is just an identifier, but it's name is indicating the later use. Is'nt it enough to call id1 and id2? maybe?

File: src/classes/RequestHandler.py [hunk 4]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
-                'modifiedDate': current_time,
-                'createdDate': current_time,
-                'metadata': resource['metadata']
+                Constants.DDB_FIELD_RESOURCE_IDENTIFIER: generated_uuid,
+                Constants.DDB_FIELD_MODIFIED_DATE: current_time,
+                Constants.DDB_FIELD_CREATED_DATE: current_time,
+                Constants.DDB_FIELD_METADATA: resource.metadata,
+                Constants.DDB_FIELD_FILES: resource.files,
+                Constants.DDB_FIELD_OWNER: resource.owner
             }
```
Comment (gregorgabriel) [inline src/classes/RequestHandler.py]: what about status and other publishing data? Is'nt there a default value for that?

File: src/classes/RequestHandler.py [hunk 5]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
+            else:
+                ddb_response = self.table.put_item(
+                    Item={
+                        Constants.DDB_FIELD_RESOURCE_IDENTIFIER: modified_resource.resource_identifier,
+                        Constants.DDB_FIELD_MODIFIED_DATE: current_time,
+                        Constants.DDB_FIELD_CREATED_DATE: previous_resource[Constants.DDB_FIELD_CREATED_DATE],
+                        Constants.DDB_FIELD_METADATA: modified_resource.metadata,
+                        Constants.DDB_FIELD_FILES: modified_resource.files,
+                        Constants.DDB_FIELD_OWNER: modified_resource.owner
+                    }
```
Comment (gregorgabriel) [inline src/classes/RequestHandler.py]: missing updateds for STATUS or publishing data

**Review by gregorgabriel — COMMENTED** [review]
Have we spend any thoughts on modifying Status and publisher information?

File: README.md [hunk 6]
```diff
@@ -4,13 +4,15 @@ Performs CRUD operations in DynamoDB
```
Comment (brinxmat) [inline README.md:4]: Should probably contain a description of the project and other information

File: src/app.py [hunk 7]
```diff
@@ -1,6 +1,10 @@
 from classes.RequestHandler import RequestHandler
+from common.constants import Constants
 
 
 def handler(event, context):
-    request_handler = RequestHandler()
-    return request_handler.handler(event, context)
+    if event is None:
```
Comment (brinxmat) [inline src/app.py:6]: We need to talk about error codes and how we get messages to frontend

File: src/classes/RequestHandler.py [hunk 8]
```diff
@@ -1,27 +1,35 @@
...
 
 
 class RequestHandler:
 
-    def __init__(self, dynamodb=None, table_name=None):
+    def __init__(self, dynamodb=None):
         if dynamodb is None:
-            self.dynamodb = boto3.resource('dynamodb')
+            self.dynamodb = boto3.resource('dynamodb', region_name=os.environ[Constants.ENV_VAR_REGION])
+            # self.dynamodb = boto3.resource('dynamodb', region_name='eu-west-1')
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:28]: Commented code does not exist

File: src/classes/RequestHandler.py [hunk 9]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
-        )
-
-        return ddb_response
+        if len(ddb_response[Constants.DDB_RESPONSE_ATTRIBUTE_NAME_ITEMS]) == 0:
+            raise ValueError('Resource with identifier ' + modified_resource.resource_identifier + ' not found')
+        else:
+            previous_resource = ddb_response[Constants.DDB_RESPONSE_ATTRIBUTE_NAME_ITEMS][0]
+            if Constants.DDB_FIELD_CREATED_DATE not in previous_resource:
+                raise ValueError(
+                    'Resource with identifier ' + modified_resource.resource_identifier + ' has no ' + Constants.DDB_FIELD_CREATED_DATE + ' in DB')
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:62]: This line is too long PEP 8

File: src/app.py [hunk 10]
```diff
@@ -1,6 +1,10 @@
 from classes.RequestHandler import RequestHandler
+from common.constants import Constants
```
Comment (brinxmat) [inline src/app.py:4]: We should implement X-Ray for tracing in every Lambda

File: src/classes/RequestHandler.py [hunk 11]
```diff
@@ -1,27 +1,35 @@
...
+    }
 
 
 class RequestHandler:
 
-    def __init__(self, dynamodb=None, table_name=None):
+    def __init__(self, dynamodb=None):
         if dynamodb is None:
-            self.dynamodb = boto3.resource('dynamodb')
+            self.dynamodb = boto3.resource('dynamodb', region_name=os.environ[Constants.ENV_VAR_REGION])
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:27]: Constant?

Comment (brinxmat) [inline src/classes/RequestHandler.py:27]: Probably set the os.environ calls as variables when app.py is initialized

File: src/classes/RequestHandler.py (hunk 4 above)
Comment (nraanes) [reply src/classes/RequestHandler.py]: Good point. We have discussed this, and such functionality must and will be implemented at a later time!

File: test/test_RequestHandler.py (hunk 3 above)
Comment (gregorgabriel) [reply test/test_RequestHandler.py]: not important

File: src/classes/RequestHandler.py [hunk 12]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
-            Item={
-                'resource_identifier': modified_resource['resource_identifier'],
-                'modifiedDate': current_time,
-                'createdDate': previous_resource['createdDate'],
-                'metadata': modified_resource['metadata']
-            }
-        )
-
-        return ddb_response
+        if len(ddb_response[Constants.DDB_RESPONSE_ATTRIBUTE_NAME_ITEMS]) == 0:
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:56]: Can this if be a standalone?

File: src/classes/RequestHandler.py [hunk 13]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
-                'headers': {'Content-Type': 'application/json'}
-            }
-        elif operation == 'MODIFY':
-            ddb_response = self.modify_resource(resource, current_time)
-            return {
-                'statusCode': 200,
-                'body': json.dumps(ddb_response),
-                'headers': {'Content-Type': 'application/json'}
-            }
+        if event is None or Constants.EVENT_BODY not in event:
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:77]: Standalone?

File: src/classes/RequestHandler.py [hunk 14]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
+            resource_dict_from_json = body.get(Constants.JSON_ATTRIBUTE_NAME_RESOURCE)
+
+            try:
+                resource = Resource.from_dict(resource_dict_from_json)
+            except TypeError as e:
+                return response(http.HTTPStatus.BAD_REQUEST, e.args[0])
+
+            current_time = arrow.utcnow().isoformat().replace('+00:00', 'Z')
+
+            if operation == Constants.OPERATION_INSERT and resource is not None:
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:91]: We check that resource is not none twice, we can extract this

File: src/classes/RequestHandler.py [hunk 15]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
+            body = json.loads(event[Constants.EVENT_BODY])
+            operation = body.get(Constants.JSON_ATTRIBUTE_NAME_OPERATION)
+            resource_dict_from_json = body.get(Constants.JSON_ATTRIBUTE_NAME_RESOURCE)
+
+            try:
+                resource = Resource.from_dict(resource_dict_from_json)
+            except TypeError as e:
+                return response(http.HTTPStatus.BAD_REQUEST, e.args[0])
+
+            current_time = arrow.utcnow().isoformat().replace('+00:00', 'Z')
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:89]: Why are we doing this replace?

File: src/classes/RequestHandler.py [hunk 16]
```diff
@@ -30,64 +38,72 @@ def get_table_connection(self):
...
+                    validate_resource(operation, resource)
+                    ddb_response = self.modify_resource(current_time, resource)
+                    ddb_response['resource_identifier'] = resource.resource_identifier
+                    return response(http.HTTPStatus.OK, json.dumps(ddb_response))
+                except ValueError as e:
+                    return response(http.HTTPStatus.BAD_REQUEST, e.args[0])
             else:
-                return int(o)
-        return super(DecimalEncoder, self).default(o)
+                return response(http.HTTPStatus.BAD_REQUEST, 'Insufficient parameters')
```
Comment (brinxmat) [inline src/classes/RequestHandler.py:109]: Constant

File: src/classes/RequestHandler.py (hunk 15 above)
Comment (nraanes) [reply src/classes/RequestHandler.py]: '2019-11-01T13:14:42.747215Z'

Comment (nraanes) [reply src/classes/RequestHandler.py:89]: '2019-11-01T13:14:42.747215Z' vs '2019-11-01T13:15:22.921610+00:00' what format do we want?

File: src/app.py (hunk 10 above)
Comment (nraanes) [reply src/app.py:4]: Made a task for this in Jira

**Pull request merged**